File: src/main/java/org/artembogomolova/demo/webapp/dao/service/UserRepoService.java
package org.artembogomolova.demo.webapp.dao.service;

import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.Comparator;
import java.util.Date;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.artembogomolova.demo.webapp.dao.repo.IAuthorityRepository;
import org.artembogomolova.demo.webapp.dao.repo.IUserRepository;
import org.artembogomolova.demo.webapp.dao.repo.IUserRoleRepository;
import org.artembogomolova.demo.webapp.domain.auth.Authority;
import org.artembogomolova.demo.webapp.domain.auth.PredefinedUserRole;
import org.artembogomolova.demo.webapp.domain.auth.Role;
import org.artembogomolova.demo.webapp.domain.auth.User;
import org.artembogomolova.demo.webapp.domain.core.Person;
import org.artembogomolova.demo.webapp.domain.core.PhysicalAddress;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

@Component
@RequiredArgsConstructor
@Slf4j
public class UserRepoService {

  private static final String PREDEFINED_ADMIN_ACCOUNT_LOGIN = "admin";
  private static final String PREDEFINED_ADMIN_ACCOUNT_COUNTRY = "Russia";
  private static final String PREDEFINED_ADMIN_ACCOUNT_POSTAL_CODE = "190000";
  private static final String PREDEFINED_ADMIN_ACCOUNT_CITY_NAME = "Saint Petersburg";
  private static final String CHANGE_IT = "changeit";

  private final IUserRepository userRepository;
  private final IUserRoleRepository userRoleRepository;
  private final IAuthorityRepository authorityRepository;

  @Transactional()
  public void createPredefinedSuperUser(PasswordEncoder passwordEncoder) {
    User result = new User();
    result.setLogin(PREDEFINED_ADMIN_ACCOUNT_LOGIN);
    result.setPassword(passwordEncoder.encode(CHANGE_IT));
    Person person = createPredefinedSuperUserPerson();
    log.info("created person: {}", person.toString());
    result.setPerson(person);
    person.setUser(result);
    Role role = userRoleRepository.findByName(PredefinedUserRole.ROLE_ADMIN.name());
    result.setRole(role);
    role.getUsers().add(result);
    result.setClientCertificateData(CHANGE_IT);
    log.info("super user role: {}", role.toString());
    role.getAuthorities().stream()
        .sorted(Comparator.comparing(Authority::getName))
        .forEach(x -> log.info("authority enabled: {}", x));
    userRepository.save(result);
  }

  private Person createPredefinedSuperUserPerson() {
    Person result = new Person();
    result.setBirthDate(new Date(LocalDateTime.now().toInstant(ZoneOffset.UTC).getEpochSecond()));
    result.setName(CHANGE_IT);
    result.setSurname(CHANGE_IT);
    result.setPatronymic(CHANGE_IT);
    result.setPhone(CHANGE_IT);
    result.setEstateAddress(buildAddress());
    return result;
  }

  private PhysicalAddress buildAddress() {
    PhysicalAddress result = new PhysicalAddress();
    result.setCountry(PREDEFINED_ADMIN_ACCOUNT_COUNTRY);
    result.setPostalCode(PREDEFINED_ADMIN_ACCOUNT_POSTAL_CODE);
    result.setCity(PREDEFINED_ADMIN_ACCOUNT_CITY_NAME);
    result.setHouse(CHANGE_IT);
    return result;
  }

  public boolean isFirstStart() {
    return userRepository.count() == 0 &&
        userRoleRepository.count() == 0 &&
        authorityRepository.count() == 0;
  }
}

File: app/src/main/kotlin/org/artembogomolova/demo/webapp/main/dao/service/UserRepoService.kt
package org.artembogomolova.demo.webapp.main.dao.service

import java.time.LocalDateTime
import java.time.ZoneOffset
import org.artembogomolova.demo.webapp.main.dao.repo.IAuthorityRepository
import org.artembogomolova.demo.webapp.main.dao.repo.IRoleRepository
import org.artembogomolova.demo.webapp.main.dao.repo.IUserRepository
import org.artembogomolova.demo.webapp.main.domain.auth.Authority
import org.artembogomolova.demo.webapp.main.domain.auth.PredefinedUserRole
import org.artembogomolova.demo.webapp.main.domain.auth.User
import org.artembogomolova.demo.webapp.main.domain.core.CountryCode
import org.artembogomolova.demo.webapp.main.domain.core.Person
import org.artembogomolova.demo.webapp.main.domain.core.PhysicalAddress
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.authentication.AnonymousAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Component
@Transactional
class UserRepoService @Autowired constructor(
    private val userRepository: IUserRepository,
    private val userRoleRepository: IRoleRepository,
    private val authorityRepository: IAuthorityRepository
) {


    fun createPredefinedSuperUser(passwordEncoder: PasswordEncoder) {
        val person = createPredefinedSuperUserPerson()
        val role = userRoleRepository.findByName(PredefinedUserRole.ADMIN.name) ?: throw IllegalStateException("role can't be null!")
        val result = User(
            login = PREDEFINED_ADMIN_ACCOUNT_LOGIN,
            password = passwordEncoder.encode(CHANGE_IT),
            clientCertificateData = CHANGE_IT,
            person = person,
            role = role
        )
        person.user = result
        println("created person: $person")
        role.users.add(result)
        println("super user role: $role")
        role.authorities.stream()
            .sorted(Comparator.comparing { obj: Authority -> obj.name!! })
            .forEach { x: Authority? -> println("authority enabled:  $x") }
        userRepository.save(result)
    }

    private fun createPredefinedSuperUserPerson(): Person = Person(
        name = CHANGE_IT,
        surname = CHANGE_IT,
        patronymic = CHANGE_IT,
        phone = EMPTY_PHONE,
        estateAddress = buildSuperUserAddress(),
        birthDate = LocalDateTime.now().toInstant(ZoneOffset.UTC).epochSecond
    )


    private fun buildSuperUserAddress(): PhysicalAddress = PhysicalAddress(
        countryCode = PREDEFINED_ADMIN_ACCOUNT_COUNTRY,
        postalCode = PREDEFINED_ACCOUNT_POSTAL_CODE,
        city = PREDEFINED_ADMIN_ACCOUNT_CITY_NAME,
        house = CHANGE_IT
    )

    fun createPredefinedGuestUser(passwordEncoder: PasswordEncoder) {
        val person = createPredefinedGuestPerson()
        val role = userRoleRepository.findByName(PredefinedUserRole.GUEST.name) ?: throw IllegalStateException("role can't be null!")
        val result = User(
            login = PREDEFINED_GUEST_ACCOUNT_LOGIN,
            password = passwordEncoder.encode(
                PREDEFINED_GUEST_ACCOUNT_LOGIN
            ),
            clientCertificateData = PREDEFINED_GUEST_ACCOUNT_LOGIN,
            role = role,
            person = person
        )
        println("created person: $person")
        println("super user role: $role")
        role.users.add(result)
        role.authorities.stream()
            .sorted(Comparator.comparing { obj: Authority -> obj.name!! })
            .forEach { x: Authority? -> println("authority enabled: $x") }
        userRepository.save(result)
    }

    private fun createPredefinedGuestPerson(): Person = Person(
        name = PREDEFINED_GUEST_ACCOUNT_LOGIN,
        surname = PREDEFINED_GUEST_ACCOUNT_LOGIN,
        patronymic = PREDEFINED_GUEST_ACCOUNT_LOGIN,
        phone = EMPTY_PHONE,
        estateAddress = buildGuestAddress(),
        birthDate = LocalDateTime.now().toInstant(ZoneOffset.UTC).epochSecond
    )


    private fun buildGuestAddress(): PhysicalAddress = PhysicalAddress(
        countryCode = PREDEFINED_GUEST_ACCOUNT_COUNTRY,
        postalCode = PREDEFINED_ACCOUNT_POSTAL_CODE,
        city = PREDEFINED_GUEST_ACCOUNT_LOGIN,
        house = PREDEFINED_GUEST_ACCOUNT_LOGIN
    )

    val isFirstStart: Boolean
        get() = userRepository.count() == 0L && userRoleRepository.count() == 0L && authorityRepository.count() == 0L
    val guestUserToken: Authentication
        get() {
            val user = userRepository.findByLogin(PREDEFINED_GUEST_ACCOUNT_LOGIN) ?: throw IllegalStateException("user can't be null!")
            val userLogin = user.login
            return AnonymousAuthenticationToken(userLogin, userLogin, user.role!!.authorities)
        }

    fun corruptedDatabase(): Boolean {
        return userRoleRepository.count() > 0 &&
                (userRepository.findByLogin(PREDEFINED_ADMIN_ACCOUNT_LOGIN) == null ||
                        userRepository.findByLogin(PREDEFINED_GUEST_ACCOUNT_LOGIN) == null)
    }

    companion object {
        const val PREDEFINED_ADMIN_ACCOUNT_LOGIN = "admin"
        private val PREDEFINED_ADMIN_ACCOUNT_COUNTRY = CountryCode.RU
        private val PREDEFINED_GUEST_ACCOUNT_COUNTRY = CountryCode.RU
        private const val PREDEFINED_ACCOUNT_POSTAL_CODE = "190000"
        private const val PREDEFINED_ADMIN_ACCOUNT_CITY_NAME = "Saint Petersburg"
        private const val CHANGE_IT = "changeit"
        const val PREDEFINED_GUEST_ACCOUNT_LOGIN = "guest"
        private const val EMPTY_PHONE = "+0-000-000-00-00"
    }
}